File: src/main/updater.js
// Auto-update: on launch, compare the CDN manifest version with what's
// installed locally. If newer, download the archive, unpack it into the game
// directory, and record the new version — no user action required.

'use strict';
const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');
const AdmZip = require('adm-zip');
const cfg = require('./config');

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// Retry a promise-returning fn a few times (core can drop connections; the CDN
// can briefly negative-cache a just-uploaded object as 403).
async function withRetry(fn, { tries = 5, delay = 1500, label = 'request' } = {}) {
  let lastErr;
  for (let i = 1; i <= tries; i++) {
    try {
      return await fn();
    } catch (e) {
      lastErr = e;
      if (i < tries) await sleep(delay);
    }
  }
  throw new Error(`${label} failed after ${tries} tries: ${lastErr && lastErr.message ? lastErr.message : lastErr}`);
}

function fetchJsonOnce(url) {
  return new Promise((resolve, reject) => {
    const lib = url.startsWith('http:') ? http : https;
    lib
      .get(url, (res) => {
        if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          return resolve(fetchJsonOnce(res.headers.location));
        }
        if (res.statusCode !== 200) {
          res.resume();
          return reject(new Error(`manifest HTTP ${res.statusCode}`));
        }
        let body = '';
        res.setEncoding('utf8');
        res.on('data', (c) => (body += c));
        res.on('end', () => {
          try {
            resolve(JSON.parse(body));
          } catch (e) {
            reject(e);
          }
        });
      })
      .on('error', reject);
  });
}

function fetchJson(url) {
  return withRetry(() => fetchJsonOnce(url), { label: 'manifest' });
}

function downloadOnce(url, dest, onProgress) {
  return new Promise((resolve, reject) => {
    const lib = url.startsWith('http:') ? http : https;
    const file = fs.createWriteStream(dest);
    lib
      .get(url, (res) => {
        if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          file.close();
          fs.rmSync(dest, { force: true });
          return resolve(downloadOnce(res.headers.location, dest, onProgress));
        }
        if (res.statusCode !== 200) {
          res.resume();
          file.close();
          fs.rmSync(dest, { force: true });
          return reject(new Error(`download HTTP ${res.statusCode}`));
        }
        const total = parseInt(res.headers['content-length'] || '0', 10);
        let got = 0;
        res.on('data', (chunk) => {
          got += chunk.length;
          if (total && onProgress) onProgress(got / total);
        });
        res.pipe(file);
        file.on('finish', () => file.close(() => resolve()));
      })
      .on('error', (err) => {
        file.close();
        fs.rmSync(dest, { force: true });
        reject(err);
      });
  });
}

function download(url, dest, onProgress) {
  return withRetry(() => downloadOnce(url, dest, onProgress), { tries: 5, delay: 2500, label: 'download' });
}

function readInstalled() {
  try {
    return JSON.parse(fs.readFileSync(cfg.versionFile, 'utf8'));
  } catch {
    return { version: null };
  }
}

function writeInstalled(version) {
  fs.mkdirSync(path.dirname(cfg.versionFile), { recursive: true });
  fs.writeFileSync(cfg.versionFile, JSON.stringify({ version, updatedAt: Date.now() }, null, 2));
}

/**
 * Ensure the newest game build is installed.
 * `emit(stage, data)` reports progress to the UI: "check" | "download" |
 * "install" | "uptodate" | "ready" | "error".
 * Returns { version } on success.
 */
async function ensureUpToDate(emit) {
  emit('check');
  let manifest;
  try {
    manifest = await fetchJson(cfg.updateManifest);
  } catch (e) {
    // Offline / CDN down: fall back to whatever is installed, if anything.
    const installed = readInstalled();
    if (installed.version) {
      emit('ready', { version: installed.version, offline: true });
      return { version: installed.version };
    }
    emit('error', { message: 'Не удалось проверить обновления и игра не установлена.' });
    throw e;
  }

  // manifest: { version: "1.2.0", url: "https://cdn.../game/petusgdps-1.2.0.zip" }
  const installed = readInstalled();
  if (installed.version === manifest.version && fs.existsSync(path.join(cfg.gameDir, cfg.exeName))) {
    emit('uptodate', { version: manifest.version });
    emit('ready', { version: manifest.version });
    return { version: manifest.version };
  }

  emit('download', { version: manifest.version, progress: 0 });
  fs.mkdirSync(cfg.gameDir, { recursive: true });
  const tmpZip = path.join(cfg.gameDir, `.update-${manifest.version}.zip`);
  await download(manifest.url, tmpZip, (p) => emit('download', { version: manifest.version, progress: p }));

  emit('install', { version: manifest.version });
  const zip = new AdmZip(tmpZip);
  zip.extractAllTo(cfg.gameDir, /* overwrite */ true);
  fs.rmSync(tmpZip, { force: true });

  writeInstalled(manifest.version);
  emit('ready', { version: manifest.version });
  return { version: manifest.version };
}

// True when a game executable is present on disk.
function isGameInstalled() {
  try {
    return fs.existsSync(path.join(cfg.gameDir, cfg.exeName));
  } catch {
    return false;
  }
}

module.exports = { ensureUpToDate, readInstalled, isGameInstalled };

File: src/Updater.cs
using System.IO.Compression;
using System.Text.Json;

namespace PetusLauncher;

record Manifest(string Version, string Url);

// Downloads + installs the game from the core manifest. Always sends a
// User-Agent (the CDN returns 403 without one) and retries transient failures.
static class Updater
{
    static readonly HttpClient Http = MakeClient();

    static HttpClient MakeClient()
    {
        var c = new HttpClient { Timeout = TimeSpan.FromMinutes(30) };
        c.DefaultRequestHeaders.UserAgent.ParseAdd("PetusLauncher/2.2");
        return c;
    }

    public static bool IsInstalled()
        => File.Exists(Path.Combine(Config.GameDir, Config.ExeName));

    public static string? InstalledVersion()
    {
        try
        {
            var doc = JsonDocument.Parse(File.ReadAllText(Config.VersionFile));
            return doc.RootElement.TryGetProperty("version", out var v) ? v.GetString() : null;
        }
        catch { return null; }
    }

    // True if a newer game build is available than what's installed.
    public static async Task<bool> UpdateAvailableAsync()
    {
        try
        {
            var m = await FetchManifestAsync();
            return IsInstalled() && InstalledVersion() != m.Version;
        }
        catch { return false; }
    }

    // Download size (Content-Length) of the current game zip — used to show the
    // install/update size without downloading. Returns 0 if unknown.
    public static async Task<long> RemoteSizeAsync()
    {
        try
        {
            var m = await FetchManifestAsync();
            using var req = new HttpRequestMessage(HttpMethod.Head, m.Url);
            using var resp = await Http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead);
            return resp.Content.Headers.ContentLength ?? 0;
        }
        catch { return 0; }
    }

    static void WriteInstalled(string version)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(Config.VersionFile)!);
        File.WriteAllText(Config.VersionFile,
            JsonSerializer.Serialize(new { version, updatedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() }));
    }

    static async Task<T> WithRetry<T>(Func<Task<T>> fn, int tries = 5, int delayMs = 2000)
    {
        Exception? last = null;
        for (int i = 1; i <= tries; i++)
        {
            try { return await fn(); }
            catch (Exception e) { last = e; if (i < tries) await Task.Delay(delayMs); }
        }
        throw new Exception($"failed after {tries} tries: {last?.Message}");
    }

    public static async Task<Manifest> FetchManifestAsync()
    {
        return await WithRetry(async () =>
        {
            var json = await Http.GetStringAsync(Config.UpdateManifest);
            var m = JsonSerializer.Deserialize<Manifest>(json,
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
            if (m == null || string.IsNullOrEmpty(m.Url)) throw new Exception("bad manifest");
            return m;
        });
    }

    // Ensure the newest build is installed. `progress` reports stage + fraction.
    // Returns the installed version. Stages: check/download/install/uptodate/ready.
    public static async Task<string> EnsureUpToDateAsync(Action<string, double> progress)
    {
        progress("check", 0);
        Manifest manifest;
        try { manifest = await FetchManifestAsync(); }
        catch
        {
            var installed = InstalledVersion();
            if (installed != null) { progress("ready", 1); return installed; }
            throw;
        }

        if (InstalledVersion() == manifest.Version && IsInstalled())
        {
            progress("uptodate", 1);
            progress("ready", 1);
            return manifest.Version;
        }

        Directory.CreateDirectory(Config.GameDir);
        var tmpZip = Path.Combine(Config.GameDir, $".update-{manifest.Version}.zip");

        await WithRetry(async () =>
        {
            progress("download", 0);
            using var resp = await Http.GetAsync(manifest.Url, HttpCompletionOption.ResponseHeadersRead);
            resp.EnsureSuccessStatusCode();
            var total = resp.Content.Headers.ContentLength ?? 0;
            await using var src = await resp.Content.ReadAsStreamAsync();
            await using var dst = File.Create(tmpZip);
            var buffer = new byte[81920];
            long got = 0;
            int read;
            while ((read = await src.ReadAsync(buffer)) > 0)
            {
                await dst.WriteAsync(buffer.AsMemory(0, read));
                got += read;
                if (total > 0) progress("download", (double)got / total);
            }
            return true;
        });

        progress("install", 0);
        // Extract over the game dir (overwrite).
        using (var archive = ZipFile.OpenRead(tmpZip))
        {
            foreach (var entry in archive.Entries)
            {
                if (string.IsNullOrEmpty(entry.Name)) continue; // directory
                var dest = Path.Combine(Config.GameDir, entry.FullName);
                Directory.CreateDirectory(Path.GetDirectoryName(dest)!);
                entry.ExtractToFile(dest, overwrite: true);
            }
        }
        try { File.Delete(tmpZip); } catch { }

        WriteInstalled(manifest.Version);
        progress("ready", 1);
        return manifest.Version;
    }
}
